### PR TITLE
Documentation comments generation enhancements.

### PR DESCRIPTION
**Pull request by zduny**
This PR includes implementation of generation of:

Resolves https://github.com/eigerco/uniffi-rs/issues/3.
Resolves https://github.com/eigerco/uniffi-rs/issues/4.

- description comments for enums,
- description comments for objects,
- description comments for object methods,
- description comments for constructors,
- descriptions for struct members,
- descriptions for enum members.

### CODE REVIEW

**Comment by zduny** [comment]
Not ready yet - discovered bugs.

Will wait for #6 to get pushed on main, rebase and then also fix arguments and return value descriptions.

**Review by eloylp — APPROVED** [review]
Awesome ! tested by:

Running a container with our [docker image](https://github.com/eigerco/uniffi-zcash-lib/blob/main/docker/Dockerfile):

```bash
docker run --rm -it -v $(pwd):/workspace:Z -w /workspace uniffi /bin/bash
```

Then
```bash
export CLASSPATH="$CLASSPATH:/opt/jna.jar"
```
Then

```bash
$ cd examples/documentation
$ cargo test
```
And it worked :tada:   . I see the comments in the bindings now. Great ! Are you planning adding this to the [general docs readme](https://github.com/eigerco/uniffi-rs/blob/main/examples/README.md) in the future ?

File: examples/documentation/tests/bindings/test_documentation.rb [hunk 1]
```diff
@@ -1,12 +1,10 @@
...
-  end
-end
+require 'test/unit'
+require 'documentation'
+
+include Test::Unit::Assertions
+
+assert_equal Documentation.hello(Documentation::Pet.new("Tom")), "Hello Tom!"
+assert_equal Documentation::Person.new("Daniel").get_name(), "Daniel"
+assert_equal Documentation.add(2, 4), 6
```
Comment (hauleth) [inline examples/documentation/tests/bindings/test_documentation.rb:10]: Hmmm, that is pretty unusual approach to Ruby testing. Why you have removed while test class?

File: uniffi_docs/src/lib.rs [hunk 2]
```diff
@@ -162,15 +171,53 @@ pub fn extract_documentation(path: &Utf8Path) -> Result<Documentation> {
...
+                    })
+                    .collect();
+
                 if let Some(description) = description {
-                    structures.insert(name, Structure { description });
+                    structures.insert(
+                        name,
+                        Structure {
+                            description,
+                            members,
```
Comment (hauleth) [inline uniffi_docs/src/lib.rs:189]: I think that calling `collect` there could improve the performance when `description == None`.

**Review by hauleth — APPROVED** [review]
Some minor questions and ideas, but overall <3

File: examples/documentation/tests/bindings/test_documentation.rb (hunk 1 above)
Comment (zduny) [reply examples/documentation/tests/bindings/test_documentation.rb:10]: They didn't work (weren't launched).
I don't know Ruby, so I don't know why, maybe `uniffi::build_foreign_language_testcases!` starts those files as scripts or something. 
I picked test structure from [here](https://github.com/eigerco/uniffi-rs/blob/main/examples/geometry/tests/bindings/test_geometry.rb).

If using class tests structure, maybe tests need to be started somehow?
Like in Python there is this section:
```py
if __name__ == '__main__':
    unittest.main()
```

Is there something like this for Ruby?

File: uniffi_bindgen/src/interface/enum_.rs [hunk 3]
```diff
@@ -88,19 +88,33 @@ use super::{APIConverter, ComponentInterface};
...
     pub(super) flat: bool,
 }
 
+impl PartialEq for Enum {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name && self.variants == other.variants && self.flat == other.flat
+    }
+}
+
+impl Eq for Enum {}
```
[the remark this answers is not in the document]
Comment (zduny) [reply uniffi_bindgen/src/interface/enum_.rs:107]: I didn't want documentation to affect the way the thing works, but yeah it got too ugly, I think I'll try simple derivation and will see if tests pass...

**Comment by zduny** [comment]
@eloylp Thanks for pointing out examples `README.md` - I'll add documentation example to it.

File: uniffi_docs/src/lib.rs (hunk 2 above)
Comment (zduny) [reply uniffi_docs/src/lib.rs:189]: Will do.

File: uniffi_docs/src/lib.rs [hunk 4]
```diff
@@ -162,15 +171,53 @@ pub fn extract_documentation(path: &Utf8Path) -> Result<Documentation> {
...
```
[the remark this answers is not in the document]
Comment (zduny) [reply uniffi_docs/src/lib.rs:217]: Will do.

File: examples/documentation/tests/bindings/test_documentation.rb (hunk 1 above)
Comment (zduny) [reply examples/documentation/tests/bindings/test_documentation.rb:10]: Done.

File: uniffi_bindgen/src/interface/enum_.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (zduny) [reply uniffi_bindgen/src/interface/enum_.rs]: Done.

File: examples/documentation/tests/bindings/test_documentation.rb (hunk 1 above)
Comment (zduny) [reply examples/documentation/tests/bindings/test_documentation.rb:10]: Leaving as is to match existing examples.

File: uniffi_docs/src/lib.rs (hunk 2 above)
Comment (zduny) [reply uniffi_docs/src/lib.rs:188]: Done

File: uniffi_docs/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (zduny) [reply uniffi_docs/src/lib.rs:215]: Done.

**Pull request merged**